### DISCUSSION TITLE
Beta Discussion

### DISCUSSION THREAD

**Discussion by allancoding** [EmulatorJS/EmulatorJS#244] [General]
### We can have a discussion here for the beta system!

> Here are some helpful things.
> 
> 1. Only ever gotten it to compile in ubuntu (I use ubuntu 20.04 in WSL)
> 2. All the shader commands are located in /menu/menu_driver.c
> 3. The default input driver (for webassembly) is /input/drivers/rwebinput_input.c. This is by far the most difficult task and is also one of the most important. We will need to basically re-write the driver.
> 4. And with the cleaning up the menu, a lot of this is going to be settings in the Makefile.emscripten, but these arent always defined correctly in the files so thats going to be a pain to go through all the errors and make sure they are in an if statement checking if the menu option is on.
> 
> The repo to the emulatorjs retroarch source: https://github.com/ethanaobrien/RetroArch
>

**Comment by allancoding** [comment]
I saw that the retroarch popup in the bottom part of the screen is there when loading states. And sometimes to play/pause button does not work. Sorry not good quality.
![Screenshot from 2022-05-16 19-44-27](https://user-images.githubusercontent.com/74841470/168711729-0e7734cf-276d-43e3-86fa-2d9fea6702dc.png)
.

**Reply by ethanaobrien** [reply]
You might want to look at the pause/play function. I had to basically re-declare it to pause the emulator instead of pausing the webassembly which is not very efficient. Without the re-declared functions, you can pause the emulator, but you cannot re-start it (meaning you have to reload the page before it starts again). I think this may be a compiler flag or a variable we need to change somewhere. Ive spent hours digging but havent been able to fix it. Maybe youll be able to find it.

And with the messages in the bottom corner, I want to take those messages out (along with the whole retroarch menu). I havent looked at this much.

And (replying to your comment from the other place)

New shaders should be pretty easy to do. I'll put them into the js files, but first could you program the [ `shader_enable`](https://github.com/ethanaobrien/RetroArch/blob/d128b5a873a7e2a4f0a526dbe1e83d12b51415d4/frontend/drivers/platform_emscripten.c#L83) function? (the function can be moved anywhere in any of the C files, for example the save function is located [here](https://github.com/ethanaobrien/RetroArch/blob/d128b5a873a7e2a4f0a526dbe1e83d12b51415d4/tasks/task_save.c#L1481). (The save function is good and does not need to be worked on)).

>and audio quality

If I am correct, this is only for nes, and Im pretty sure the old versions of the core support this. This isnt much of a concern for me (as the new beta cores use much less cpu and sound quality (based on my research) doesnt really change the performance.

>and color palet

Im pretty sure we have this too (for gameboy only I think). What systems are this for?

>And I agree with the menu button idea.

Ive just opened #245 

See [this](https://github.com/ethanaobrien/emulatorjs/blob/main/data/README.md#building) for compiling instructions. I recommend using ubuntu 20.04 using the Windows Subsystem for linux feature. It is available in windows 10 and 11 and is very easy to use (in my opinion). Also, If you need a lightweight dev web server, I have [this](https://github.com/ethanaobrien/C-server). Its in development still, but it works.

**Reply by allancoding** [reply]
But do you see the popup in green in the screenshot? Loded from slot #0 is that ok. Mean I don't care but it's not super clean with emulatojs.

**Reply by ethanaobrien** [reply]
Sorry, just added some edits. Thats part of the retroarch menu and I would like to take out the menu entirely, it would save a lot of space and be much cleaner. I havent really looked on how to remove the messages yet.

**Reply by allancoding** [reply]
Yes because I was looking through the code in menu folder. And saw that part of the code.

**Reply by ethanaobrien** [reply]
Well yeah, A lot of that can be removed (just make the functions empty, deleting them can cause errors)

**Reply by allancoding** [reply]
Ok and I just multiboot ubuntu desktop so I just use VIsual Studio Code and https://github.com/ritwickdey/vscode-live-server to use local host live edit. It might take me sometime but I will be working on it in a couple of days because I am almost done with school!! Then I will have all summer! Thank you for your help.

**Reply by ethanaobrien** [reply]
No problem, and thank you for your help!

**Reply by allancoding** [reply]
PS: I had to install: https://emscripten.org/docs/getting_started/downloads.html#installation-instructions To build.

**Reply by ethanaobrien** [reply]
Oh yeah, forgot to add that to the read me. Sorry

**Reply by allancoding** [reply]
I can not build GearBoy or snes9x

**Comment by allancoding** [comment]
Where is the function `toggleMainLoop` because on beta it throws an error mainLoop is not a function.
```
if (_0x55627a.coreVer === 2) {
     var mainLoop = _0x4d7024.Module.cwrap('toggleMainLoop', 'null', ['number']);
          _0x4d7024.Module.resumeMainLoop = function() {
               mainLoop(1);
          }
          _0x4d7024.Module.pauseMainLoop = function() {
               mainLoop(0);
          }
}
```
UPDATE!!: It works find on the nes but not the snes. Did you just compile that only on the nes and not the snes? https://github.com/ethanaobrien/emulatorjs/commit/2fcc7bd47bde6e65e68f7ea5f03301293b3d4a0f

**Reply by ethanaobrien** [reply]
>Did you just compile that only on the nes and not the snes

Yes, that is correct. I didnt have the time to compile everything

**Comment by allancoding** [comment]
How do you compile the .data files from the wasm and .js file?

**Reply by ethanaobrien** [reply]
7zip together the wasm and js files. You do not need to do renaming. (must be in the `.7z` format). Then rename to `nes-wasm.data` or whatever your the name would be for the core you are building

**Comment by ethanaobrien** [comment]
Did it give you an error of any kind?
